MyListLibrary/MyList.cs

```diff
@@ -0,0 +1,343 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace MyListLibrary
+{
+    public class MyList<T> : IEnumerable<T>, IMyList<T> where T : IComparable<T>
+    {
+        private const int DEFAULT_SIZE = 4;
+        private const double INCREAMENTOR = 1.33;
+
+        private T[] array;
+        public int Count { get; set; }
+        public int Capacity { get => array.Length; }
+
+        public MyList()
+        {
+            this.array = new T[DEFAULT_SIZE];
+        }
+
+        public MyList(T value)
+        {
+            this.array = new T[DEFAULT_SIZE];
+            array[0] = value;
+            ++Count;
+        }
+
+        public MyList(T[] array)
+        {
+            this.array = array;
+            Count += array.Length;
+        }
+
+        public T this[int index]
+        {
+            get
+            {
+                if (index >= Count)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+                return array[index];
+            }
+        }
+
+        public void AddFirst(T value)
+        {
+            AddByIndex(0, value);
+        }
+
+        public void AddLast(T value)
+        {
+            AddByIndex(Count, value);
+        }
+
+        public void AddByIndex(int index, T value)
+        {
+            if (index < 0 || index > Count)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            if (IsFull())
+            {
+                Resize();
+            }
+
+            for (int i = Count - 1; i > index; i--)
+            {
+                array[i + 1] = array[i];
+            }
+
+            array[index] = value;
+
+            ++Count;
+        }
+
+        public void AddElementsFirst(IEnumerable<T> list)
+        {
+            AddElementsByIndex(0, list);
+        }
+
+        public void AddElementsLast(IEnumerable<T> list)
+        {
+            AddElementsByIndex(Count, list);
+        }
+
+        public void AddElementsByIndex(int index, IEnumerable<T> list)
+        {
+            if (index < 0 || index > Count)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            MyList<T> arr = (MyList<T>)list;
```

Comment:
No use child classes. It ruins interface advantages.

Reply:
fixed